### PR TITLE
specification: Device interface assignment

### PR DESCRIPTION
**Pull request by sameo**

### CODE REVIEW

File: specification/07-theory_operations.adoc [hunk 1]
```diff
@@ -525,15 +525,62 @@ host supervisor domain manager may directly assign a TDI to a TVM, through the
...
+2. The TSM generates and sends a TDISP `LOCK_INTERFACE_REQUEST` message to the
+   DSM, which is encapsulated into a vendor-defined SPDM request.
+3. The TSM receives the TDISP `LOCK_INTERFACE_RESPONSE` message, which contains
+   a device-generated nonce. The TSM stores the nonce.
+4. The TSM programs the DMA mappings into the trusted IOMMU space, in order for
+   the TDI inbound transactions to map into the TVM address space. The IOMMU
+   mappings are programmed but not validated. They will be validated only once
+   the TDI moves to the TDISP `RUN` state, i.e. when it accepts the TDI into
+   its TCB through the `sbi_covg_run_interface()` `COVG` ABI.
+5. For each of the assigned TDI MMIO range, the host supervisor domain manager
```
Comment (sameo) [inline specification/07-theory_operations.adoc:559]: This should be done before calling `sbi_covh_add_interface()`

File: specification/07-theory_operations.adoc [hunk 2]
```diff
@@ -525,15 +525,62 @@ host supervisor domain manager may directly assign a TDI to a TVM, through the
...
+
+==== Assignment Flow
+
+Assigning an interface to a TVM goes through the following steps:
+
+1. The host supervisor domain manager calls the `sbi_covh_add_interface()`
+   `COVH` function to initiate the TDI assignment. The TSM rejects that request
+   if a secured SPDM session is not established with the DSM or if a stream IDE
+   link is not set up.
+2. The TSM generates and sends a TDISP `LOCK_INTERFACE_REQUEST` message to the
```
Comment (jyao1) [inline specification/07-theory_operations.adoc:550]: The overall flow is good.

Question: Do we want to list all TDISP sequence?

I think the TSM needs to send `GET_TDISP_VERSION` and `GET_TDISP_CAPABILITIES`, before `LOCK_INTERFACE_REQUEST`.

File: specification/07-theory_operations.adoc [hunk 3]
```diff
@@ -525,15 +525,62 @@ host supervisor domain manager may directly assign a TDI to a TVM, through the
...
+   transaction will not trigger MMIO page faults. Here as well, the TSM does
+   not enable the above mappings until the TVM moves the TDI to the TDISP `RUN`
+   state.
+6. Through regular discovery mechanisms (ACPI, PCI bus scanning), the TVM
+   detects the assigned TDI. It is important to note that the TDI configuration
+   space, including its BARs, is emulated by the host security domain manager.
+   The TDI MMIO ranges are mapped into the TVM address space by the the host
+   security domain manager itself.
+7. Before using the TDI, the TVM must
+   xref:tdi-acceptation[accept it into its TCB]. Moreover, the TVM must not use
```
Comment (jyao1) [inline specification/07-theory_operations.adoc:571]: The current tdi-acceptation section does not have any flow. Such as when GET_MEASUREMENT is sent, when GET_INTERFACE_REPORT is sent.  

Do we need to address that?

File: specification/07-theory_operations.adoc (hunk 1 above)
Comment (sameo) [reply specification/07-theory_operations.adoc]: Fixed

File: specification/07-theory_operations.adoc (hunk 2 above)
Comment (sameo) [reply specification/07-theory_operations.adoc]: > The overall flow is good.
> 
> Question: Do we want to list all TDISP sequence?

I think we should, yes.

> I think the TSM needs to send `GET_TDISP_VERSION` and `GET_TDISP_CAPABILITIES`, before `LOCK_INTERFACE_REQUEST`.

I added those now.

File: specification/07-theory_operations.adoc (hunk 3 above)
Comment (sameo) [reply specification/07-theory_operations.adoc:575]: I added a sequence diagram for the verification and the acceptation flows. Let me know how it looks.

File: specification/09-coveio_abi.adoc [hunk 4]
```diff
@@ -105,3 +127,17 @@ struct sbiret sbi_covg_map_interface_mmio(unsigned long device_if_id
                                           unsigned long hpa_offset,
                                           unsigned long size);
 ----
+
+[#sbi_covg_run_interface()]
+===== Run Device Interface (FID TBD)
```
Comment (jyao1) [inline specification/09-coveio_abi.adoc:132]: to match `sbi_covg_stop_interface`, maybe we can call it `sbi_covg_start_interface` ?

File: specification/07-theory_operations.adoc [hunk 5]
```diff
@@ -512,29 +512,147 @@ VMM ->> Device: Enable IDE for the selected stream
...
+process for binding an interface to a TVM, and one if its most critical steps:
+the TVM decision of accepting or rejecting the TDI into its TCB.
+
+==== Binding Flow
+
+Binding an interface and a TVM together goes through the following steps:
+
+1. For the binding process to succeed, the host supervisor domain manager must
+   first pass each of the bound TDI MMIO ranges to the TVM. It does so by
+   calling the `sbi_covh_add_tvm_device_region()` `COVH` ABI. The TSM services
```
Comment (jyao1) [inline specification/07-theory_operations.adoc:542]: I do not find `sbi_covh_add_tvm_device_region` definition in chap 9.

Do you mean `sbi_covh_connect_device` or `sbi_covh_bind_interface` ?

Comment (sameo) [reply specification/07-theory_operations.adoc:542]: > I do not find `sbi_covh_add_tvm_device_region` definition in chap 9.

Should be fixed now.

> Do you mean `sbi_covh_connect_device` or `sbi_covh_bind_interface` ?

I mean `sbi_covh_bind_interface`. Does that make sense to you?

Comment (jyao1) [reply specification/07-theory_operations.adoc:542]: yes

File: specification/09-coveio_abi.adoc (hunk 4 above)
Comment (jyao1) [reply specification/09-coveio_abi.adoc]: I did a global search, still find `sbi_covg_run_interface` ...

Comment (sameo) [reply specification/09-coveio_abi.adoc]: Ah, sorry. That *should* be fixed now...

File: specification/09-coveio_abi.adoc [hunk 6]
```diff
@@ -52,8 +52,35 @@ identified by `stream_id`.
 struct sbiret sbi_covh_disconnect_device()(unsigned long device_id);
 -----
 
-==== Device Interface Assignment
+==== TVM Memory Management
 
+[#sbi_covh_add_tvm_device_region()]
```
Comment (jyao1) [inline specification/09-coveio_abi.adoc:57]: Do we need corresponding `remove` style API?

Comment (sameo) [reply specification/09-coveio_abi.adoc:57]: Would it make sense to e.g. remove one TDI MMIO region from the TVM, while keeping the TDI bound to the TVM? Wouldn't the TDI be mostly useless for the TVM at that point? If that's so, I think we should not allow the TVM to do that and only have it reclaim the TDI by unbinding it (`sbi_covh_unbind_interface`) if it wants to unmap those regions from the TVM.
What do you think?

Comment (jyao1) [reply specification/09-coveio_abi.adoc:57]: It is more for the secure clean up.

I expect that the TVM may request the VMM to unbind the device. Then the VMM needs to **remove the device related MMIO and DMA**, from the TVM domain, by calling TSM interface.

Later, the TVM can verify if VMM has done that correctly, by calling TSM interface to check.

**Comment by jyao1** [comment]
@sameo , please also resolve conflict.

File: specification/09-coveio_abi.adoc (hunk 6 above)
Comment (sameo) [reply specification/09-coveio_abi.adoc:57]: Makes a lot of sense to me, thanks. I added a `sbi_covh_reclaim_tvm_interface_region` API.

**Comment by sameo** [comment]
> @sameo , please also resolve conflict.

All conflicts fixed now.

**Review by jyao1 — APPROVED** [review]

**Pull request merged**